examples/rvc-app/rvc-common/src/rvc-device.cpp

```diff
@@ -0,0 +1,107 @@
+
+    switch (NewMode)
+    {
+    case RvcRunMode::ModeMapping: {
+        // change to mapping only allowed from the Idle state.
+        if (currentMode != RvcRunMode::ModeIdle)
+        {
+            response.status = to_underlying(ModeBase::StatusCode::kGenericFailure);
+            response.statusText.SetValue(chip::CharSpan::fromCharString("Change to the mapping mode is only allowed from idle"));
+            return;
+        }
+
+        mRunModeInstance.UpdateCurrentMode(NewMode);
+        // Set operational state to running
+        ChipError err =
+            mOperationalStateInstance.SetOperationalState(to_underlying(OperationalState::OperationalStateEnum::kRunning));
+        if (err != CHIP_NO_ERROR)
+        {
+            response.status = to_underlying(ModeBase::StatusCode::kGenericFailure);
+            response.statusText.SetValue(chip::CharSpan::fromCharString(err.AsString()));
+            return;
+        }
+        break;
+    }
+    case RvcRunMode::ModeCleaning: {
+        mRunModeInstance.UpdateCurrentMode(NewMode);
+        // Set operational state to running
+        ChipError err =
+            mOperationalStateInstance.SetOperationalState(to_underlying(OperationalState::OperationalStateEnum::kRunning));
+        if (err != CHIP_NO_ERROR)
+        {
+            response.status = to_underlying(ModeBase::StatusCode::kGenericFailure);
+            response.statusText.SetValue(chip::CharSpan::fromCharString(err.AsString()));
+            return;
+        }
+    }
+    break;
+    case RvcRunMode::ModeIdle: {
+        mRunModeInstance.UpdateCurrentMode(NewMode);
+        // Set operational state to running
+        ChipError err =
+            mOperationalStateInstance.SetOperationalState(to_underlying(OperationalState::OperationalStateEnum::kStopped));
+        if (err != CHIP_NO_ERROR)
+        {
+            response.status = to_underlying(ModeBase::StatusCode::kGenericFailure);
+            response.statusText.SetValue(chip::CharSpan::fromCharString(err.AsString()));
+            return;
+        }
+    }
+    }
+
+    response.status = to_underlying(ModeBase::StatusCode::kSuccess);
+}
+
+void RvcDevice::HandleRvcCleanChangeToMode(uint8_t NewMode, ModeBase::Commands::ChangeToModeResponse::Type & response)
+{
+    uint8_t rvcRunCurrentMode = mRunModeInstance.GetCurrentMode();
+
+    if (rvcRunCurrentMode != RvcRunMode::ModeIdle)
+    {
+        response.status = to_underlying(RvcCleanMode::StatusCode::kCleaningInProgress);
+        response.statusText.SetValue(chip::CharSpan::fromCharString("Change of the cleaning mode is only allowed in Idle."));
+        return;
+    }
+
+    response.status = to_underlying(ModeBase::StatusCode::kSuccess);
+}
+
+void RvcDevice::HandleOpStatePauseCallback(Clusters::OperationalState::GenericOperationalError & err)
+{
+    auto error = mOperationalStateInstance.SetOperationalState(to_underlying(OperationalState::OperationalStateEnum::kPaused));
```

Comment:
Speaking of API.. maybe there should be a template version of SetOperationalState that only allows enum-typed args and does the to_underlying thing, to make the API nicer to use?  Similar for the other APIs here that require using to_underlying.

This should be a followup, not more changes in this PR.

Reply:
Is it possible to make these kind of changes after the SVE freeze? I can create some issues to keep track of this work if so.

Reply:
> Is it possible to make these kind of changes after the SVE freeze?

Absolutely.  As long as they don't affect wire-observable behavior...

Reply:
I have added [this](https://github.com/project-chip/connectedhomeip/issues/29085) issue.